### PR TITLE
DAMD-63: Implement FluxTable

### PR DESCRIPTION
**Pull request by PaulPrice**

### CODE REVIEW

File: datamodel.txt [hunk 1]
```diff
@@ -360,32 +360,41 @@ The path would be
...
+               filterName                              [STRING]
+               fiberMag                                [FLOAT]
+HDU #3 SKY         Sky flux in same units as FLUX    	       [FLOAT]	      NROW
+HDU #4 OBSERVATIONS    Binary table                            [FITS BINARY TABLE] NOBS
+               Columns for:
+               identity of the observation             [STRING]
+               fiberId                                 [32-bit INT]
+               nominal PFI position                    [FLOAT]*2
+               actual PFI position                     [FLOAT]*2
+HDU #5 COVAR       Near-diagonal part of HDU 2's covariance    [FLOAT]        NROW*3
```
Comment (hassanxp) [inline datamodel.txt]: As the HDUs have been renumbered, the description referencing HDU 2 is incorrect. It should now be HDU 7 I believe.

File: datamodel.txt [hunk 2]
```diff
@@ -365,32 +365,46 @@ The path would be
...
+               filterName                              [STRING]
+               fiberMag                                [FLOAT]
+HDU #4 SKY         Sky flux in same units as FLUX             [FLOAT]        NROW
+HDU #5 COVAR       Near-diagonal part of FLUX's covariance    [FLOAT]        NROW*3
+HDU #6 COVAR2      Low-resolution non-sparse estimate covariance [FLOAT]      NCOARSE*NCOARSE
+HDU #7 OBSERVATIONS    Binary table                            [FITS BINARY TABLE] NOBS
+               Columns for:
+               identity of the observation             [STRING]
+               fiberId                                 [32-bit INT]
+               nominal PFI position                    [FLOAT]*2
```
Comment (RobertLuptonTheGood) [inline datamodel.txt]: Units?  mm in focal plane or (ra, dec) are possibilities.  I think I prefer mm as it's closer to what we actually know (although it still requires MCS distortion mapping to be correct)

File: datamodel.txt [hunk 3]
```diff
@@ -365,32 +365,46 @@ The path would be
 The file will have several HDUs:
 
 HDU #0 PDU
-HDU #1 FLUX        Flux in units of nJy     	   	       [FLOAT]        NROW
-HDU #2 FLUXTBL     Binary table                                [FITS BINARY TABLE] NROW
-    		   Columns for:
+HDU #1 FLUX        Flux in units of nJy                       [FLOAT]        NROW
+HDU #2 MASK        Pixel mask                                 [32-bit INT]   NROW
```
Comment (RobertLuptonTheGood) [inline datamodel.txt]: As you've already moved HDUs around is it worth moving the COVAR up too?

File: datamodel.txt [hunk 4]
```diff
@@ -365,32 +365,46 @@ The path would be
...
+               fiberId                                 [32-bit INT]
+               nominal PFI position                    [FLOAT]*2
+               actual PFI position                     [FLOAT]*2
+HDU #8 FLUXTABLE   Binary table                                [FITS BINARY TABLE] NOBS*NROW
+               Columns for:
 		       wavelength in units of nm (vacuum)      [FLOAT]
-		       intensity in units of nJy	       [FLOAT]
-		       intensity error same units as intensity [FLOAT]
+		       intensity in units of nJy	           [FLOAT]
+		       variance in same units as intensity     [FLOAT]
```
Comment (RobertLuptonTheGood) [inline datamodel.txt]: Why did you change this from error to variance?  If so, it's in nJy^2 not the same as intensity

File: python/pfs/datamodel/pfsSpectra.py [hunk 5]
```diff
@@ -86,6 +91,24 @@ def __str__(self):
         """Stringify"""
         return "%s{%d spectra of length %d}" % (self.__class__.__name__, self.numSpectra, self.length)
 
+    def __imul__(self, rhs):
+        """In-place multiplication"""
+        with np.errstate(invalid="ignore"):
+            self.flux *= rhs
+            self.sky *= rhs
+            for ii in range(3):
+                self.covar[:, ii, :] *= rhs
```
Comment (RobertLuptonTheGood) [inline python/pfs/datamodel/pfsSpectra.py]: Isn't this `*= rhs**2`?

File: python/pfs/datamodel/pfsSpectra.py [hunk 6]
```diff
@@ -86,6 +91,24 @@ def __str__(self):
...
+        return self
+
+    def __itruediv__(self, rhs):
+        """In-place division"""
+        with np.errstate(invalid="ignore", divide="ignore"):
+            self.flux /= rhs
+            self.sky /= rhs
+            for ii in range(3):
+                self.covar[:, ii, :] /= rhs
+        return self
```
Comment (RobertLuptonTheGood) [inline python/pfs/datamodel/pfsSpectra.py]: Isn't this `./= rhs**2`?

File: python/pfs/datamodel/interpolate.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    try:
+        # kind="previous" and kind="next" requires scipy v1.1.0 or later
+        return np.bitwise_or(impl("previous"), impl("next"))
+    except NotImplementedError:
+        # Grow the bad pixels by 1
+        array = impl("nearest")
+        result = array.copy()
+        result[1:] |= array[:-1]
+        result[:-1] |= array[1:]
+        return result
```
Comment (RobertLuptonTheGood) [inline python/pfs/datamodel/interpolate.py]: This whole implementation is a place-holder from RHL's febrile brain.  Now it's factored out into its own function (thanks) it should be moved into DRP and, if needs be, included in the 1-D simulator by `git submodules` magic, like `datamodel` itself.

File: python/pfs/datamodel/pfsSpectra.py (hunk 5 above)
Comment (RobertLuptonTheGood) [inline python/pfs/datamodel/pfsSpectra.py]: What about `covar2`?

File: python/pfs/datamodel/pfsSpectrum.py [hunk 8]
```diff
@@ -355,3 +414,24 @@ def plot(self, plotSky=True, plotErrors=True, ignorePixelMask=0x0, show=True):
         if show:
             figure.show()
         return figure, axes
+
```
Comment (RobertLuptonTheGood) [inline python/pfs/datamodel/pfsSpectrum.py]: I think this needs to be factored out along with interpolate.py.   It's a place-holder, and this sort of implementation belongs squarely in the pipeline code

File: python/pfs/datamodel/interpolate.py (hunk 7 above)
Comment (PaulPrice) [reply python/pfs/datamodel/interpolate.py]: It's used for convenience methods in datamodel classes. This is bending the rules about algorithmic content in the datamodel package, but I was thinking users might want to be able to interpolate spectra or the flux table as part of their scientific investigations, and I thought the convenience of having such a basic functionality was worth it. What do you think?

File: datamodel.txt (hunk 2 above)
Comment (PaulPrice) [reply datamodel.txt]: Shall I put "units TBD"? I like mm too, but we're really just going to pass through whatever we're given from the instrument.

File: datamodel.txt (hunk 3 above)
Comment (PaulPrice) [reply datamodel.txt]: The order of the HDUs is governed at least in part due to some inheritance going on. There is a simplified version of `PfsSpectrum` called `PfsSimpleSpectrum` which we use for reference spectra, and it contains flux and mask and object, but no sky or variance/covariance (because it's from a model). The sky and variance extensions come after the object extension because it reflects the inheritance (base class first, and then add on customisations).

File: python/pfs/datamodel/pfsSpectra.py (hunk 5 above)
Comment (PaulPrice) [reply python/pfs/datamodel/pfsSpectra.py]: I'm not sure what to do with `covar2`, probably because I haven't grokked what it's for. I guess I've always thought that it's a feature that could be implemented after everything else is working nicely.
In this particular case, for scalars inputs the behaviour might be straightforward, but this method is often called with vector arguments, which have a very different shape, so it's not clear to me how to deal with `covar2`.

File: datamodel.txt (hunk 2 above)
Comment (RobertLuptonTheGood) [reply datamodel.txt]: That's hard.  I think I'd go with mm for now.  Certainly not ra, dec

File: python/pfs/datamodel/interpolate.py (hunk 7 above)
Comment (RobertLuptonTheGood) [reply python/pfs/datamodel/interpolate.py]: I don't think we want naive algorithms in `datamodel` that people might use.  We should make sure that they can use the real functions (at least from jupyter).  I'm fine with being careful to make them as purely pythonic as possible

File: python/pfs/datamodel/pfsSpectra.py (hunk 5 above)
Comment (RobertLuptonTheGood) [reply python/pfs/datamodel/pfsSpectra.py]: It captures over-all errors in spectrophotometry.  I am OK with not handling it here for now (for the reasons given) if we assert on it's being not filled out yet

File: datamodel.txt (hunk 2 above)
Comment (PaulPrice) [reply datamodel.txt]: Ah, a few other places in `datamodel.txt` use microns. Adopting that.

File: python/pfs/datamodel/interpolate.py (hunk 7 above)
Comment (PaulPrice) [reply python/pfs/datamodel/interpolate.py]: OK. I propose doing that on a separate ticket so that we can get this in.

Comment (PaulPrice) [reply python/pfs/datamodel/interpolate.py]: [PIPE2D-466](https://pfspipe.ipmu.jp/jira/browse/PIPE2D-466).
